[v16] build: Ensure we target the min version supported on Mac builds

camscale opened the pull request:
Ensure relevant flags are passed to the Go, rust and C compilers (xcode)
when building the Teleport binaries to target the minimum version of
macOS we support. Currently that is 10.15 (Catalina) on x86_64 and 11
(Big Sur) on arm64 (the first supported Apple silicon macOS release).

Note that we set the minimum to 10.15 in the build, but it will be
updated to 11 by the compilers/tools when building for ARM64.

Care is taken with rust builds not to set this min os globally as we
also use rust to target web assembly in the web UI and the flags are
invalid there.

Backport: https://github.com/gravitational/teleport/pull/43231
Backport: https://github.com/gravitational/teleport/pull/43996
Link: https://goteleport.com/docs/installation/#operating-system-support
Issue: https://github.com/gravitational/teleport/issues/32601
Companion: https://github.com/gravitational/teleport.e/pull/4612

---

I'dd add an `e` ref update to include the companion PR once that gets merged.

camscale committed:
[v16] build: Ensure we target the min version supported on Mac builds

Ensure relevant flags are passed to the Go, rust and C compilers (xcode)
when building the Teleport binaries to target the minimum version of
macOS we support. Currently that is 10.15 (Catalina) on x86_64 and 11
(Big Sur) on arm64 (the first supported Apple silicon macOS release).

Note that we set the minimum to 10.15 in the build, but it will be
updated to 11 by the compilers/tools when building for ARM64.

Care is taken with rust builds not to set this min os globally as we
also use rust to target web assembly in the web UI and the flags are
invalid there.

Link: https://goteleport.com/docs/installation/#operating-system-support

espadolini reviewed (commented):
Do we need to apply the same treatment to the fdpass-teleport build?

camscale commented:
> Do we need to apply the same treatment to the fdpass-teleport build?

It looks like it. Building it on my machine and running `vtool -show-build build/fdpass-teleport` reports minos as 14.0. This change will need to be made to master, which I'll do now, and add to this backport.

camscale committed:
[v16] build: Simplify the vars that target minimum macOS version

Simplify how we tell rust/xcode/go which minimum version of macOS to
target. Xcode defines the environment variable
`MACOSX_DEPLOYMENT_TARGET` that contains the version of macOS to target.
llvm has adopted this too, so cargo/rust supports this, meaning we no
longer need to use `RUSTFLAGS` with a linker flag. This is much simpler
as we need to be careful not to set this flag when targetting wasm for
the UI (ironrdp). When building Go with CGO, it uses the system linker,
which is xcode, so we can rely on the `MACOSX_DEPLOYMENT_TARGET` var
there too, allowing us to get rid of the `CGO_LDFLAGS` var. We do still
need to set a linker flag on `CGO_CFLAGS` though, from testing.

This also ensures that `fdpass-teleport` also meets the macOS minimum
version requirements. It is nice that we no longer need special handling
on each rust target to do this.

camscale committed:
Update `e` ref for macOS minos changes
